### PR TITLE
fix(backup/restore): fixes backup and restore with DROP operations (GRAPHQL-735)

### PR DESCRIPTION
**Pull request by abhimanyusinghgaur**
Fixes GRAPHQL-735.
See [RFC](https://discuss.dgraph.io/t/backup-and-restore-with-operations-like-drop-all/11218) for more details.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6844)

 
 

Docs Preview: [Dgraph Preview](https://dgraph-c5e2a2b433-107582.surge.sh)

### CODE REVIEW

File: worker/backup_processor.go [hunk 1]
```diff
@@ -310,3 +323,49 @@ func writeKVList(list *bpb.KVList, w io.Writer) error {
...
+		case "DROP_DATA":
+			dropOp.DropOp = pb.DropOperation_DATA
+		case "DROP_ATTR":
+			dropOp.DropOp = pb.DropOperation_ATTR
+			dropOp.DropValue = dropInfo[1]
+		}
+		return dropOp, nil
+	default:
+		// getting more than one values for a non-list predicate is an error
+		return nil, errors.Errorf("found multiple values for dgraph.drop.op")
```
Comment (jarifibrahim) [inline worker/backup_processor.go:369]: You might want to add the vals here so that it helps in debugging.

File: worker/backup_processor.go [hunk 2]
```diff
@@ -310,3 +323,49 @@ func writeKVList(list *bpb.KVList, w io.Writer) error {
...
+	}
+	switch len(vals) {
+	case 0:
+		// do nothing, it means this one was deleted with S * * deletion.
+		// So, no need to consider it.
+		return nil, nil
+	case 1:
+		val, ok := vals[0].Value.([]byte)
+		if !ok {
+			return nil, errors.Errorf("cannot convert value of dgraph.drop.op to byte array")
```
Comment (jarifibrahim) [inline worker/backup_processor.go:345]: Add the actual type of the interface to the error. `reflect.TypeOf(...)` would give you this information.

File: worker/backup_ee.go [hunk 3]
```diff
@@ -171,12 +171,16 @@ func ProcessBackupRequest(ctx context.Context, req *pb.BackupRequest, forceFull
...
 	errCh := make(chan error, len(state.Groups))
+	var dropOperations []*pb.DropOperation
 	for _, gid := range groups {
 		br := proto.Clone(req).(*pb.BackupRequest)
 		br.GroupId = gid
 		br.Predicates = predMap[gid]
 		go func(req *pb.BackupRequest) {
-			_, err := BackupGroup(ctx, req)
+			res, err := BackupGroup(ctx, req)
+			if res != nil && len(res.DropOperations) > 0 {
```
Comment (jarifibrahim) [inline worker/backup_ee.go:181]: You don't need to check the length. See https://play.golang.org/p/odlQYbwcQL2

File: worker/backup_ee.go [hunk 4]
```diff
@@ -171,12 +171,16 @@ func ProcessBackupRequest(ctx context.Context, req *pb.BackupRequest, forceFull
...
+				dropOperations = append(dropOperations, res.DropOperations...)
```
Comment (jarifibrahim) [inline worker/backup_ee.go:182]: This append happens in multiple go routines. You'll have a race condition.

File: worker/backup_ee.go [hunk 5]
```diff
@@ -188,7 +192,7 @@ func ProcessBackupRequest(ctx context.Context, req *pb.BackupRequest, forceFull
 		}
 	}
 
-	m := Manifest{Since: req.ReadTs, Groups: predMap}
+	m := Manifest{Since: req.ReadTs, Groups: predMap, DropOperations: dropOperations}
```
Comment (jarifibrahim) [inline worker/backup_ee.go:195]: Should you append the new `dropOperations` to the existing backups? What if you had incremental backups?

**Review by jarifibrahim — COMMENTED** [review]
Got some comments.

File: worker/backup_processor.go (hunk 1 above)
Comment (abhimanyusinghgaur) [reply worker/backup_processor.go]: Done

File: worker/backup_processor.go (hunk 2 above)
Comment (abhimanyusinghgaur) [reply worker/backup_processor.go]: Done

File: worker/backup_ee.go (hunk 3 above)
Comment (abhimanyusinghgaur) [reply worker/backup_ee.go]: Done

File: worker/backup_ee.go (hunk 4 above)
Comment (abhimanyusinghgaur) [reply worker/backup_ee.go]: Fixed, thanks!

File: worker/backup_ee.go (hunk 5 above)
Comment (abhimanyusinghgaur) [reply worker/backup_ee.go:201]: No, each backup whether full or incremental creates a new manifest file. So, here it is writing only to the new manifest file, no old files are modified.

**Review by pawanrawal — APPROVED** [review]
> Reviewed 15 of 26 files at r1, 6 of 13 files at r2, 15 of 15 files at r3.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6844)* status: all files reviewed, 13 unresolved discussions (waiting on @abhimanyusinghgaur, @jarifibrahim, @manishrjain, @MichaelJCompton, and @vvbalaji-dgraph)

___
*[edgraph/server.go, line 339 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6844#-MLgalRT0XVcaQpr9i49:-MLgalRT0XVcaQpr9i4A:bmdw73v) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5e2a2b4336e47786508c4e8a22362712cd23c22/edgraph/server.go#L339)):*
> ```Go
> 		}
> 
> 		// insert a helper record for backup & restore, indicating that drop_all was done
> ```

Add a comment that if the server crashes before storing this, that would be an issue. 
___
*[protos/pb.proto, line 613 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6844#-MLgbLvL9N40VBZj-syb:-MLgbLvL9N40VBZj-syc:b-uypq95) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5e2a2b4336e47786508c4e8a22362712cd23c22/protos/pb.proto#L613)):*
> ```protobuf
> 	}
> 	DropOp drop_op    = 1;
> 	string drop_value = 2;
> ```

Add a comment that this would only be present when DropOp is ATTR.
___
*[systest/online-restore/online_restore_test.go, line 587 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6844#-MLgdJRz7z1fP24AEPyh:-MLgdJRz7z1fP24AEPyi:b-my8j0d) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5e2a2b4336e47786508c4e8a22362712cd23c22/systest/online-restore/online_restore_test.go#L587)):*
> ```Go
> 	// remove backup directories to make sure this test doesn't leave anything behind
> 	// TODO: This is having some problem on TeamCity
> 	//cleanupDirs(t, localFsDirs)
> ```

Can you create a JIRA ticket to look into it when you come back?
___
*[worker/backup_ee.go, line 173 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6844#-MLge1Sl9CNS_Siy-3hw:-MLge1Sl9CNS_Siy-3hx:b-74ikxr) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5e2a2b4336e47786508c4e8a22362712cd23c22/worker/backup_ee.go#L173)):*
> ```Go
> 	defer cancel()
> 
> 	resOrErrCh := make(chan interface{}, len(state.Groups))
> ```

type backupRes {
  res *BackupResponse
  err error
}

chan backupRes
___
*[worker/backup_ee.go, line 190 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6844#-MLgeSkb2llQFzH57FLU:-MLgeSkb2llQFzH57FLV:b-em4p3v) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5e2a2b4336e47786508c4e8a22362712cd23c22/worker/backup_ee.go#L190)):*
> ```Go
> 	var dropOperations []*pb.DropOperation
> 	for range groups {
> 		if resOrErr := <-resOrErrCh; resOrErr != nil {
> ```

res.err != nil 
___
*[worker/backup_processor.go, line 136 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6844#-MLgenjh6t_bnCwZM6ff:-MLgenjh6t_bnCwZM6fg:b-l5bd98) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5e2a2b4336e47786508c4e8a22362712cd23c22/worker/backup_processor.go#L136)):*
> ```Go
> 		tl := pr.threads[itr.ThreadId]
> 		tl.alloc = stream.Allocator(itr.ThreadId)
> 		kvList, dropOp, err := tl.toBackupList(key, itr)
> ```

if err != nil {
  return nil, err
}
// append
// return kvList, nil
___
*[worker/backup_processor.go, line 346 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6844#-MLgfyfE1LL3vbycTLJh:-MLgfyfFAjh0Xr0MibhP:b-dcu1f3) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5e2a2b4336e47786508c4e8a22362712cd23c22/worker/backup_processor.go#L346)):*
> ```Go
> 		val, ok := vals[0].Value.([]byte)
> 		if !ok {
> 			return nil, errors.Errorf("cannot convert value of dgraph.drop.op to byte array, "+
> ```

Also, add the value that is stored.
___
*[x/keys.go, line 560 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6844#-MLggjDu5JW1xEIznf2v:-MLggjDu5JW1xEIznf2w:b-jiwz5) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5e2a2b4336e47786508c4e8a22362712cd23c22/x/keys.go#L560)):*
> ```Go
> 	"dgraph.graphql.schema":            {},
> 	"dgraph.cors":                      {},
> 	"dgraph.drop.op":                   {},
> ```

Add a TODO that we should rename this map.

**Review by abhimanyusinghgaur — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6844)* status: 30 of 36 files reviewed, 13 unresolved discussions (waiting on @jarifibrahim, @manishrjain, @MichaelJCompton, @pawanrawal, and @vvbalaji-dgraph)

___
*[edgraph/server.go, line 339 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6844#-MLgalRT0XVcaQpr9i49:-MLgvtno6eiFik6yWcKZ:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5e2a2b4336e47786508c4e8a22362712cd23c22/edgraph/server.go#L339)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Add a comment that if the server crashes before storing this, that would be an issue. 
</blockquote></details>

Done.
___
*[protos/pb.proto, line 613 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6844#-MLgbLvL9N40VBZj-syb:-MLgvvAW4CMWCRIcXqxG:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5e2a2b4336e47786508c4e8a22362712cd23c22/protos/pb.proto#L613)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Add a comment that this would only be present when DropOp is ATTR.
</blockquote></details>

Done.
___
*[systest/online-restore/online_restore_test.go, line 587 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6844#-MLgdJRz7z1fP24AEPyh:-MLgvw182gUihHYVKXur:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5e2a2b4336e47786508c4e8a22362712cd23c22/systest/online-restore/online_restore_test.go#L587)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Can you create a JIRA ticket to look into it when you come back?
</blockquote></details>

Done.
___
*[worker/backup_ee.go, line 173 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6844#-MLge1Sl9CNS_Siy-3hw:-MLgvxjI1HoZijPPtPXd:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5e2a2b4336e47786508c4e8a22362712cd23c22/worker/backup_ee.go#L173)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

type backupRes {
  res *BackupResponse
  err error
}

chan backupRes
</blockquote></details>

Done.
___
*[worker/backup_ee.go, line 190 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6844#-MLgeSkb2llQFzH57FLU:-MLgw1ql5HCtbyRZjAYE:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5e2a2b4336e47786508c4e8a22362712cd23c22/worker/backup_ee.go#L190)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

res.err != nil 
</blockquote></details>

Done.
___
*[worker/backup_processor.go, line 136 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6844#-MLgenjh6t_bnCwZM6ff:-MLgw8oP22gMJkLdNxOp:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5e2a2b4336e47786508c4e8a22362712cd23c22/worker/backup_processor.go#L136)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

if err != nil {
  return nil, err
}
// append
// return kvList, nil
</blockquote></details>

Done.
___
*[worker/backup_processor.go, line 346 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6844#-MLgfyfE1LL3vbycTLJh:-MLgwCQC1O62aUm1m54i:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5e2a2b4336e47786508c4e8a22362712cd23c22/worker/backup_processor.go#L346)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Also, add the value that is stored.
</blockquote></details>

Done.
___
*[x/keys.go, line 560 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6844#-MLggjDu5JW1xEIznf2v:-MLgwHKD9yzOumrXPD_U:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/c5e2a2b4336e47786508c4e8a22362712cd23c22/x/keys.go#L560)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Add a TODO that we should rename this map.
</blockquote></details>

Done.

**Pull request merged**